### PR TITLE
Add metadata based compile time logger purging

### LINKED ISSUE #7095: Allow compile time purging to happen per application/module
Requires #7046.

Mailing list discussion: https://groups.google.com/d/msgid/elixir-lang-core/606029ea-9855-44c6-ba94-69113ffed45c%40googlegroups.com?utm_medium=email&utm_source=footer

### PR DESCRIPTION
**Pull request by josevalim**
Closes #7095

### CODE REVIEW

File: lib/logger/lib/logger.ex [hunk 1]
```diff
@@ -88,19 +88,35 @@ defmodule Logger do
       #{perform_cleanup()}")` as in the example `perform_cleanup/0` won't be
       executed if the `:compile_time_purge_level` is `:info` or higher.
 
+    * `:compile_time_purge_matching` - purges *at compilation time* all calls
+      that matches the given metadata. This configuration expects a list of
```
Comment (fertapric) [inline lib/logger/lib/logger.ex]: I think it should be "all calls that *match* the given metadata"

File: lib/logger/lib/logger.ex [hunk 2]
```diff
@@ -88,19 +88,35 @@ defmodule Logger do
...
+      keyword lists. Each keyword list contains a metadata key and the matching
+      value that should be purged. Remember that if you want to put purge log
+      calls from a dependency, the dependency must be recompiled
```
Comment (fertapric) [inline lib/logger/lib/logger.ex]: By looking at the other paragraphs, I think a dot at the end is missing here.

File: lib/logger/lib/logger.ex [hunk 3]
```diff
@@ -88,19 +88,35 @@ defmodule Logger do
       #{perform_cleanup()}")` as in the example `perform_cleanup/0` won't be
       executed if the `:compile_time_purge_level` is `:info` or higher.
 
+    * `:compile_time_purge_matching` - purges *at compilation time* all calls
+      that matches the given metadata. This configuration expects a list of
+      keyword lists. Each keyword list contains a metadata key and the matching
+      value that should be purged. Remember that if you want to put purge log
```
Comment (fertapric) [inline lib/logger/lib/logger.ex]: Maybe "Remember that if you want to purge log calls from a dependency, the dependency must be recompiled."?

**Comment by josevalim** [comment]
Thank you @fertapric! Can you also please review #7768 if you haven't yet?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**